spec/github_helper_spec.rb

```diff
@@ -171,4 +133,187 @@ def mock_comment(body: '<!-- REUSE_ID: test-id --> Test', user_id: 1234)
       instance_double('Comment', id: 1234, body: body, user: instance_double('User', id: user_id))
     end
   end
+
+  describe '#initialize' do
+    let(:client) do
+      instance_double(
+        Octokit::Client,
+        user: instance_double('User', name: 'test'),
+        'auto_paginate=': nil
+      )
+    end
+
+    before do
+      allow(Octokit::Client).to receive(:new).and_return(client)
+    end
+
+    it 'properly passes the token all the way down to the Octokit::Client' do
+      expect(Octokit::Client).to receive(:new).with(access_token: 'Fake-GitHubToken-123')
+      described_class.new(github_token: 'Fake-GitHubToken-123')
+    end
```

Comment:
What do you think of simplifying this to 

```suggestion
    it 'properly passes the token all the way down to the Octokit::Client' do
      allow(Octokit::Client).to receive(:new).and_return(client)
      expect(Octokit::Client).to receive(:new).with(access_token: 'Fake-GitHubToken-123')
      described_class.new(github_token: 'Fake-GitHubToken-123')
    end
```

Reply:
Also, testing that a certain method on a certain object is called with a certain parameter makes the test heavily dependent on the implementation details of the system under test as well as the object it checks.

There's nothing bad about this approach, but it can backfire because we want to be able to change the code's implementation (i.e. refactor) many times and use the tests to verify our changes haven't broken the code's behavior. Because, at the end of the day, it's the behavior that really matters.

I think this test in particular is a bit of an edge case, because what we are testing is inherently an implementation detail: Whether `Octokit::Client` received the correct token. So, I think it's okay to keep the test as is.

Just curious, did you happen to find other possible ways to verify the behavior of the token having been passed to `Octokit::Client`? (_Apologies if this has already been discussed. There are a lot of resolved conversations in this PR and I haven't expanded them all_).

Reply:
Hey @mokagio, thanks for your feedback! 
I really appreciate it! 😁 

>Just curious, did you happen to find other possible ways to verify the behavior of the token having been passed to Octokit::Client?

To be honest, yes I have some other options that I was considering to follow here to validate the token instead of expect `Octokit::Client::new` having its value. I list them here with the tradeoffs that I consider them to have, and really would like yours and @AliSoftware inputs about them.

## Reinstate the empty/nil validation at the token

Before I start this PR, on the `github_helper` we have the `github_token` function that tries to get the token from the Env variables and [raised an error if is not present](https://github.com/wordpress-mobile/release-toolkit/blob/6d474fe9faed59ac8259ff0866982c434c4de0d9/lib/fastlane/plugin/wpmreleasetoolkit/helper/github_helper.rb#L11-L20). We can have the same validation in place at our `initialize` function as the first thing, something like this:

```ruby
token || UI.user_error!('Please provide a GitHub authentication token')
```

With this, we can validate the presence of the token, no matter what we're going to do with it later and, as a bonus, we remove the `Octokit::Client::new` expectations as it is now on this PR. The drawback here is that we "lose" our assurance that the `Octokit::Client` receives the token that we pass at the helper, and no other thing in the place instead.

Although this will be also validated by the ConfigItens at the actions and is a very edge case having a GithubHelper without the token present, with this double check we prevent any changes on that requirement of having the token, in the future.

## Test the API Call that Oktokit does at the initialization

This other option exists, TBH I do not like it as much as the first as I see no value in testing things inside 3rd-party libraries. 

Looking at the library source code, at the moment that the `Octokit::Client` is initialized with the access token, it does a request to the GitHub API under the hood, passing that token as one of the Headers of that request. I imagine that is due to validate if the token is indeed valid before proceeding with anything. 

```
 GET https://api.github.com/user with headers {'Accept'=>'application/vnd.github.v3+json', 'Accept-Encoding'=>'gzip;q=1.0,deflate;q=0.6,identity;q=0.3', 'Authorization'=>'token Fake-GitHubToken-123', 'Content-Type'=>'application/json', 'User-Agent'=>'Octokit Ruby Gem 4.21.0'}
```

So, we could stub that call and check if the token that we passed is indeed in that request. However, as I said before, this is an implementation detail at the `Octokit::Client` side, which is most certainly also covered by their internal unit tests. 

Although we remove the dependency of knowing if the `Octokit::Client::new` has our correct token, we also introduce another dependency to an internal thing that the library does and thus can be changed in the future also being a dangerous drawback.

Reply:
Hey @mokagio, thanks for your feedback! 
I really appreciate it! 😁 

>Just curious, did you happen to find other possible ways to verify the behavior of the token having been passed to Octokit::Client?

To be honest, yes I have some other options that I was considering to follow here to validate the token instead of expect `Octokit::Client::new` having its value. I list them here with the tradeoffs that I consider them to have, and really would like yours and @AliSoftware inputs about them.

## Reinstate the empty/nil validation at the token

Before I start this PR, on the `github_helper` we have the `github_token` function that tries to get the token from the Env variables and [raised an error if is not present](https://github.com/wordpress-mobile/release-toolkit/blob/6d474fe9faed59ac8259ff0866982c434c4de0d9/lib/fastlane/plugin/wpmreleasetoolkit/helper/github_helper.rb#L11-L20). We can have the same validation in place at our `initialize` function as the first thing, something like this:

```ruby
token || UI.user_error!('Please provide a GitHub authentication token')
```

With this, we can validate the presence of the token, no matter what we're going to do with it later and, as a bonus, we remove the `Octokit::Client::new` expectations as it is now on this PR. The drawback here is that we "lose" our assurance that the `Octokit::Client` receives the token that we pass at the helper, and no other thing in the place instead.

Although this will be also validated by the ConfigItens at the actions and is a very edge case having a GithubHelper without the token present, with this double check we prevent any changes on that requirement of having the token, in the future.

## Test the API Call that Oktokit does at the initialization

This other option exists, TBH I do not like it as much as the first as I see no value in testing things inside 3rd-party libraries. 

Looking at the library source code, at the moment that the `Octokit::Client` is initialized with the access token, it does a request to the GitHub API under the hood, passing that token as one of the Headers of that request. I imagine that is due to validate if the token is indeed valid before proceeding with anything. 

```
 GET https://api.github.com/user with headers {'Accept'=>'application/vnd.github.v3+json', 'Accept-Encoding'=>'gzip;q=1.0,deflate;q=0.6,identity;q=0.3', 'Authorization'=>'token Fake-GitHubToken-123', 'Content-Type'=>'application/json', 'User-Agent'=>'Octokit Ruby Gem 4.21.0'}
```

So, we could stub that call and check if the token that we passed is indeed in that request. However, as I said before, this is an implementation detail at the `Octokit::Client` side, which is most certainly also covered by their internal unit tests. 

Although we remove the dependency of knowing if the `Octokit::Client::new` has our correct token, we also introduce another dependency to an internal thing that the library does and thus can be changed in the future also being a dangerous drawback.

Reply:
Thank you for thinking through some options @raafaelima 😄 As I mentioned in my earlier comment, I think the code as is has acceptable tradeoff and wouldn't endorse changing it, unless someone can come with an obviously better option, of course. Still, it's fun to toss ideas around and consider options.

> Reinstate the empty/nil validation at the token

I think this test would be less valuable than what we have already. Checking that `GithubHelper` errors if the token is `nil` doesn't prevent callers from passing `nil` to it, and, as you point out, it wouldn't tell us whether `Octokit::Client` gets that token.

We might be better served by [static types checking](https://github.com/wordpress-mobile/release-toolkit/issues/201) for this. _One day..._

> Test the API Call that Oktokit does at the initialization
>
> [...] we also introduce another dependency to an internal thing that the library does and thus can be changed in the future also being a dangerous drawback.

This is interesting. I kinda like the idea to check whether an HTTP request to the GitHub API is made with the given token. This effect should stay the same if for some reason we changed how we interface with the API in the future—we always end up needing to talk to the GitHub APIs at some point.

On the other hand, a great advantage of delegating the GitHub API interaction implementation to OctoKit is that we don't need to worry about these lower level details. Maybe accepting a test a bit more towards the implementation details side of the spectrum is worth not having to worry about all the GitHub API side of thing.